### PR TITLE
feat: add support for deserialized response

### PR DESCRIPTION
**Pull request by alvin-reyes**
# Changes

support for deserialized response based on the requirement https://github.com/filecoin-project/boost/issues/1531

#### Set value to `all`
```
echo Starting booster-http...
exec booster-http run --serve-gateway=all --api-lid=$LID_API_INFO  --api-fullnode=$FULLNODE_API_INFO --api-storage=$MINER_API_INFO --tracing
```
```
2023-08-20 19:15:01 2023-08-20T10:15:01.190Z    INFO    booster lib/api.go:38   Using full node API at ws://lotus:1234/rpc/v1
2023-08-20 19:15:01 2023-08-20T10:15:01.193Z    INFO    booster booster-http/run.go:156 Tracing exporter enabled
2023-08-20 19:15:01 2023-08-20T10:15:01.193Z    INFO    booster lib/api.go:64   Using storage API at ws://lotus-miner:2345/rpc/v0
2023-08-20 19:15:01 2023-08-20T10:15:01.194Z    INFO    booster lib/api.go:106  Miner address: t01000
2023-08-20 19:15:18 2023-08-20T10:15:18.354Z    INFO    booster booster-http/run.go:217 Starting booster-http node on listen address 0.0.0.0 and port 7777 with base path ''
2023-08-20 19:15:18 2023-08-20T10:15:18.355Z    INFO    booster booster-http/run.go:224 serving IPFS gateway at /ipfs/ (serving files)
2023-08-20 19:15:18 2023-08-20T10:15:18.355Z    INFO    booster booster-http/run.go:226 serving raw pieces at /piece/
```
image

#### Set value to `verifiable`
```
echo Starting booster-http...
exec booster-http run --serve-gateway=verifiable --api-lid=$LID_API_INFO  --api-fullnode=$FULLNODE_API_INFO --api-storage=$MINER_API_INFO --tracing
```
```
2023-08-21 19:58:36 http://boost:8042
2023-08-21 19:58:36 Starting booster-http...
2023-08-21 19:58:36 2023-08-21T10:58:36.942Z    INFO    booster lib/api.go:38   Using full node API at ws://lotus:1234/rpc/v1
2023-08-21 19:58:36 2023-08-21T10:58:36.950Z    INFO    booster booster-http/run.go:155 Tracing exporter enabled
2023-08-21 19:58:36 2023-08-21T10:58:36.950Z    INFO    booster lib/api.go:64   Using storage API at ws://lotus-miner:2345/rpc/v0
2023-08-21 19:58:36 2023-08-21T10:58:36.951Z    INFO    booster lib/api.go:106  Miner address: t01000
2023-08-21 19:58:46 2023-08-21T10:58:46.378Z    INFO    booster booster-http/run.go:216 Starting booster-http node on listen address 0.0.0.0 and port 7777 with base path ''
2023-08-21 19:58:46 2023-08-21T10:58:46.379Z    INFO    booster booster-http/run.go:223 serving IPFS gateway at /ipfs/ (serving blocks, CARs)
2023-08-21 19:58:46 2023-08-21T10:58:46.379Z    INFO    booster booster-http/run.go:225 serving raw pieces at /piece/
```
image

#### set value to `none`
```
echo Starting booster-http...
exec booster-http run --serve-gateway=none --api-lid=$LID_API_INFO  --api-fullnode=$FULLNODE_API_INFO --api-storage=$MINER_API_INFO --tracing
```
```
2023-08-21 19:58:36 2023-08-21T10:58:36.942Z    INFO    booster lib/api.go:38   Using full node API at ws://lotus:1234/rpc/v1
2023-08-21 19:58:36 2023-08-21T10:58:36.950Z    INFO    booster booster-http/run.go:155 Tracing exporter enabled
2023-08-21 19:58:36 2023-08-21T10:58:36.950Z    INFO    booster lib/api.go:64   Using storage API at ws://lotus-miner:2345/rpc/v0
2023-08-21 19:58:36 2023-08-21T10:58:36.951Z    INFO    booster lib/api.go:106  Miner address: t01000
2023-08-21 19:58:46 2023-08-21T10:58:46.378Z    INFO    booster booster-http/run.go:216 Starting booster-http node on listen address 0.0.0.0 and port 7777 with base path ''
2023-08-21 19:58:46 2023-08-21T10:58:46.379Z    INFO    booster booster-http/run.go:223 serving IPFS gateway at /ipfs/ (serving blocks, CARs)
2023-08-21 19:58:46 2023-08-21T10:58:46.379Z    INFO    booster booster-http/run.go:225 serving raw pieces at /piece/
```
image

### made a online unverified deal for a simple car file
```
boost -vv deal --verified=false --provider=t01000 --commp=baga6ea4seaqdcoodqrjm65fkx5bstoh5aqe57zx7rbulpnfyp477b6d4vtyespq --car-size=104 --piece-size=128 --http-url=https://radiant.edge.estuary.tech/gw/bafybeieo26tuorkzzh5p347ucintyxgaqichmga3pv4euinokvyvemj5ya --payload-cid=bafkqagcineqee33pon2cyid3gexc4njxpuqhi2lnmvzqu --storage-price=1000000000

2023-08-22T08:19:14.983Z        DEBUG   boost   boost/deal_cmd.go:279   about to submit deal proposal   {"uuid": "d011c703-7bca-4a43-8629-0f40c737a283"}
sent deal proposal
  deal uuid: d011c703-7bca-4a43-8629-0f40c737a283
  storage provider: t01000
  client wallet: t3quyq27bdclsakjo5usrhejle3wch46fbcysvtk2vy5c2gwkqmtgco6jgeerxbjqot7v5o5rnnoan57b75pxa
  payload cid: bafkqagcineqee33pon2cyid3gexc4njxpuqhi2lnmvzqu
  url: https://radiant.edge.estuary.tech/gw/bafybeieo26tuorkzzh5p347ucintyxgaqichmga3pv4euinokvyvemj5ya
  commp: baga6ea4seaqdcoodqrjm65fkx5bstoh5aqe57zx7rbulpnfyp477b6d4vtyespq
  start epoch: 6096
  end epoch: 524496
  provider collateral: 0
```
image

### Ready to publish deal
image

#### Look by CID
image

### Configure booster-http to set serve-gateway=verifiable
image

### Configure booster-http to set serve-gateway=none
image

### CODE REVIEW

File: cmd/booster-http/run.go [hunk 1]
```diff
@@ -76,20 +76,10 @@ var runCmd = &cli.Command{
...
-			Usage: "serve CAR files with the ipfs gateway API",
-			Value: true,
-		},
-		&cli.BoolFlag{
-			Name:  "serve-files",
-			Usage: "serve original files (eg jpg, mov) with the ipfs gateway API",
-			Value: false,
+		&cli.StringFlag{
+			Name:  "serve-gateway",
+			Usage: "serve deserialized responses with the ipfs gateway API (values: 'verifiable','all','none')",
```
Comment (dirkmc) [inline cmd/booster-http/run.go:81]: I'd suggest we add a Description to StringFlag that describes how the behaviour of booster-http changes when specifying each of the options: verifiable, all and none

File: cmd/booster-http/gateway_handler.go [hunk 2]
```diff
@@ -14,19 +14,26 @@ type gatewayHandler struct {
 	supportedFormats map[string]struct{}
 }
 
-func newGatewayHandler(gw *gateway.BlocksBackend, supportedFormats []string) http.Handler {
+func newGatewayHandler(gw *gateway.BlocksBackend, supportedFormats []string, serveGateway string) http.Handler {
```
Comment (dirkmc) [inline cmd/booster-http/gateway_handler.go:17]: I don't believe we need the supportFormats parameter any more.
I think we can also get rid of most of the code in gateway_handler.go

**Review by dirkmc — COMMENTED** [review]
Looks good so far 👍 
I left a couple of suggestions for code changes and descriptions.
Please run a few tests with these new parameters and paste the results in the PR comments.

**Comment by dirkmc** [comment]
One nit: I'd also suggest using a code formatting tool for consistent whitespace

File: cmd/booster-http/run.go (hunk 1 above)
Comment (alvin-reyes) [reply cmd/booster-http/run.go]: ack. changes made on the last commit.

File: cmd/booster-http/gateway_handler.go (hunk 2 above)
Comment (alvin-reyes) [reply cmd/booster-http/gateway_handler.go]: ack. changes made on the latest commit.

File: cmd/booster-http/run.go [hunk 3]
```diff
@@ -262,32 +254,25 @@ var runCmd = &cli.Command{
 	},
 }
 
-func parseSupportedResponseFormats(cctx *cli.Context) []string {
+func parseSupportedResponseFormats(cctx *cli.Context) ([]string) {
```
Comment (dirkmc) [inline cmd/booster-http/run.go:257]: I don't think we need parseSupportedResponseFormats anymore. We can probably replace the call to it with code that checks that serve-gateway is set to verifiable / all / none

File: cmd/booster-http/gateway_handler.go [hunk 4]
```diff
@@ -14,21 +14,23 @@ type gatewayHandler struct {
...
 
-	fmtsMap := make(map[string]struct{}, len(supportedFormats))
-	for _, f := range supportedFormats {
-		fmtsMap[f] = struct{}{}
+	if serveGateway == "none" {
+		return &gatewayHandler{}
+	} else if serveGateway == "all" {
+		deserializedResponse = true
+	}else if serveGateway == "verifiable" {
+		deserializedResponse = false
```
Comment (dirkmc) [inline cmd/booster-http/gateway_handler.go:27]:
```suggestion
	if serveGateway == "none" {
		return &gatewayHandler{}
	}
	if serveGateway == "all" {
		deserializedResponse = true
	} else if serveGateway == "verifiable" {
		deserializedResponse = false
```

File: cmd/booster-http/gateway_handler.go [hunk 5]
```diff
@@ -14,21 +14,23 @@ type gatewayHandler struct {
...
 	}
-
 	return &gatewayHandler{
 		gwh: gateway.NewHandler(gateway.Config{
 			Headers:               headers,
-			DeserializedResponses: true,
+			DeserializedResponses: deserializedResponse,
 		}, gw),
-		supportedFormats: fmtsMap,
```
Comment (dirkmc) [inline cmd/booster-http/gateway_handler.go:31]: If `supportedFormats` is nil, it looks like the code will panic when `h.supportedFormats` is accessed on line 44 (inside `ServeHTTP`)

**Comment by dirkmc** [comment]
Before the next review, please run a few tests with these new parameters and paste the results in the PR comments.

File: cmd/booster-http/run.go (hunk 3 above)
Comment (alvin-reyes) [reply cmd/booster-http/run.go:257]: we can change this method to return a bool. If it's none, then it'll return false.

```
func shouldEnableIPFSGateway(cctx *cli.Context) (bool) {
	switch cctx.String("serve-gateway") {
	case "verifiable":
	case "all":
		return true
	}
	return false
}
```

File: cmd/booster-http/gateway_handler.go (hunk 5 above)
Comment (alvin-reyes) [reply cmd/booster-http/gateway_handler.go:31]: changes this on the latest commit.

File: cmd/booster-http/run.go (hunk 3 above)
Comment (alvin-reyes) [reply cmd/booster-http/run.go]: added the new function and replace the code with it.

**Comment by dirkmc** [comment]
Before the next review, please manually run booster-http with the serve-gateway parameter set to each possible value, and paste the results in the PR comments.

**Comment by alvin-reyes** [comment]
@dirkmc I added the verification on the PR description above - link here https://github.com/filecoin-project/boost/pull/1625#issue-1851126872

**Comment by dirkmc** [comment]
Thanks @alvin-reyes

Could you please also test out the functionality itself. So for each of all / verifiable / none try to retrieve a cid and verify that booster-http gives the appropriate response. So for `all` it should allow you to retrieve any cid. For `verifiable` it should only return verifiable responses. For none it should return an error message.

You can read more about the exact behaviour [in the issue description](https://github.com/filecoin-project/boost/issues/1531) and in the [Enable serving of files](https://boost.filecoin.io/retrieving-data-from-filecoin/serving-files-with-booster-http#enable-serving-of-files) of the booster-http documentation.

I also left a couple more suggestions for changes to the code.

**Comment by alvin-reyes** [comment]
I have added the verification on the PR description (sent you a message as well @dirkmc).

**Comment by LexLuthr** [comment]
Frisbii takes care of this in booster-http.

**Pull request closed without merging**